File: sdkcreationmw/sdkruntimes/MIDP/nei/src/com/symbian/tools/j2me/sei/emulator/session/commands/GetPortCommand.java
/*
* Copyright (c) 2003 - 2004 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description: 
*
*/

package com.symbian.tools.j2me.sei.emulator.session.commands;


import java.io.*;

public class GetPortCommand extends SessionCommand{

  //
  // Constants
  //

  /**
   * Get port commmand ID
   */
  public static final int GET_PORT_CMD_ID = 7;


  //
  // Members
  //

  /**
   * Free port
   */
  private int[] iFreePorts;



  //
  // Life cycle
  //

  /**
   * Constructor
   *
   * @aNum number of free ports
   */
  public GetPortCommand(int aNum){
    super(GET_PORT_CMD_ID);
    iFreePorts = new int[aNum];
  }


  //
  // Operations
  //


  /**
   * Get free ports
   *
   * @return array of free ports numbers
   */
  public int[] getFreePorts(){
    return iFreePorts;
  }


  /**
   * Dump the object to the output stream
   *
   * @param aDos Data output stream
   *
   * @throws IOException if some error occured
   */
  public void dump(DataOutputStream aDos) throws IOException{
    super.dump(aDos);
    aDos.writeInt(iFreePorts.length);
  }



  /**
   * Load the object from the stream
   *
   * @param aDis data input stream
   * @throws IOException if some error occured
   */
  public void load(DataInputStream aDis) throws IOException{
    for(int i = 0; i < iFreePorts.length; i++){
      iFreePorts[i] = aDis.readInt();
    }
  }

}

File: sdkcreationmw/sdkruntimes/MIDP/DebugAgent/debugagent_orig/src/seislave/commands/GetPortCommand.cpp
/*
* Copyright (c) 2004 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description: 
*
*/


#include "SeiCommands.h"

#include <es_sock.h>
#include <in_sock.h>



CGetPortCommand* CGetPortCommand::NewL(
    CDriver* aDriver, 
    CSocketReader* aReader,
    CSocketWriter* aWriter)
    {
    CGetPortCommand* object = new (ELeave) CGetPortCommand(aDriver, aReader, aWriter);
    CleanupStack::PushL(object);
    object->ConstructL();
    CleanupStack::Pop(object);
    return object;
    }

CGetPortCommand::CGetPortCommand(
    CDriver* aDriver, 
    CSocketReader* aReader,
    CSocketWriter* aWriter)
    : CSeiCommand(aDriver, aReader, aWriter)
    {
    }

CGetPortCommand::~CGetPortCommand()
    {
	iFreePorts.Close();
    }

void CGetPortCommand::ConstructL()
    {
    }

void CGetPortCommand::DoReadL()
    {
	iFreePortsNum = iReader->ReadIntL();
    }

void CGetPortCommand::DoExecuteL()
    {
	TInt errorCode = KErrNone;
	RSocketServ	socketServer;
	errorCode = socketServer.Connect();
	if(errorCode  == KErrNone)
		{
	    RSocket listeningSocket;
		TSockAddr localAddress;
		//TODO: add retrying to find ports
		for(int i = 0; errorCode == KErrNone && i < iFreePortsNum; i++)
			{
			TInt port = iDriver->NextRandomPort();
			listeningSocket.Open(socketServer, KAfInet, KSockStream, KProtocolInetTcp);
			localAddress.SetPort(port); 
			errorCode = listeningSocket.Bind(localAddress);
			if(errorCode == KErrNone)
				{
				iFreePorts.Append(port);
				}
			listeningSocket.Close();
			}
		}
	socketServer.Close();
	User::LeaveIfError(errorCode);
    }

void CGetPortCommand::DoWriteL()
    {
		for(int i = 0; i < iFreePortsNum; i++)
		{
			iWriter->WriteIntL(iFreePorts[i]);
		}
    }



